[run-api-tests] Limit log lines for tests

JonWBedard opened the pull request:
#### 0df1e37069766b6f679f83adafcedba26481c255
<pre>
[run-api-tests] Limit log lines for tests
<a href="https://bugs.webkit.org/show_bug.cgi?id=263446">https://bugs.webkit.org/show_bug.cgi?id=263446</a>
rdar://117265047

Reviewed by Aakash Jain.

Limit the number of logging lines a single API test can
output. By default, limit a given test to 250 lines.

* Tools/Scripts/webkitpy/api_tests/runner.py:
(setup_shard): Pass log_limit.
(Runner.__init__): Ditto.
(Runner.run): Ditto.
(_Worker): Ditto.
(_Worker.setup): Ditto.
(_Worker.__init__): Ditto.
(_Worker._run_single_test): Count number of log lines, and
fail test if we exceed the log limit.
(_Worker.run): Ditto.

Canonical link: <a href="https://commits.webkit.org/269576@main">https://commits.webkit.org/269576@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/acb8b5ee29740f69356c30af31054699ec29fb46

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22967 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/1042 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24075 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24881 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/21257 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/23231 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/2274 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23508 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/22135 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23209 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/516 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19920 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25738 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/23154 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/404 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20801 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/27002 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/20778 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21060 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/24861 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/494 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/18297 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/409 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/902 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2908 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/671 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Tools/Scripts/webkitpy/api_tests/runner.py:
```diff
@@ -210,6 +214,17 @@ def _run_single_test(self, binary_name, test):
+                    stderr_line = self.EXCEEDED_LOG_LINE_MESSAGE.format(self.log_limit)
+                    stderr_buffer += stderr_line
+                    _log.error(stderr_line[:-1])
+                    server_process.stop()
```
Out of curiosity, is this a SIGKILL or something more polite?

JonWBedard replied:
It's `SIGKILL`, but I think that's appropriate in this case.

emw-apple reviewed (commented):
Implementation seems fine, though I’m not sure how I feel yet about this as a policy change. Two questions:

This will presumably cause failures in trunk over time, because of stdio messages emitted as side effects of the test. So, in effect, this destabilizes API tests. Is that worth it?

What am I supposed to do if I’m the unlucky one who adds the 250th line of output to a test?

aj062 reviewed (approved):
I am fine with landing it for fixing the emergency for now.

JonWBedard committed:
[run-api-tests] Limit log lines for tests
https://bugs.webkit.org/show_bug.cgi?id=263446
rdar://117265047

Reviewed by Aakash Jain.

Limit the number of logging lines a single API test can
output. By default, limit a given test to 250 lines.

* Tools/Scripts/webkitpy/api_tests/runner.py:
(setup_shard): Pass log_limit.
(Runner.__init__): Ditto.
(Runner.run): Ditto.
(_Worker): Ditto.
(_Worker.setup): Ditto.
(_Worker.__init__): Ditto.
(_Worker._run_single_test): Count number of log lines, and
fail test if we exceed the log limit.
(_Worker.run): Ditto.

Canonical link: https://commits.webkit.org/269576@main
